### PR TITLE
Env var optional defaults

### PR DESCRIPTION
**Pull request by CharlieC3**
Modified NPM_CONFIG_PRODUCTION and NODE_MODULES_CACHE to use the current set value of the variable, and to be set to "true" if the variables are not already set.

### CODE REVIEW

**Comment by CharlieC3** [comment]
Hi @hunterloftis,
In regards to my comment posted here:
https://github.com/heroku/heroku-buildpack-nodejs/pull/147

In the current state, I'm unable to pass in a value to this variables as defined here in the README:
https://github.com/heroku/heroku-buildpack-nodejs#enable-or-disable-devdependencies-installation

This change would allow defaults to be set if the variables do not already contain a value passed in.

File: bin/compile [hunk 1]
```diff
@@ -27,8 +27,8 @@ unset GIT_DIR
 trap build_failed ERR
 
 # Load config vars into environment; start with defaults
-export NPM_CONFIG_PRODUCTION=true
-export NODE_MODULES_CACHE=true
+export NPM_CONFIG_PRODUCTION=${NPM_CONFIG_PRODUCTION:-true}
+export NODE_MODULES_CACHE=${NODE_MODULES_CACHE:-true}
 export_env_dir $env_dir
```
Comment (hunterloftis) [inline bin/compile]: The line that loads the env vars (`export_env_dir`) is below these exports, so user config would never be loaded.

Let's get to the bottom of why you're not seeing `NPM_CONFIG_PRODUCTION` change your app's behavior; if that's the case, then these tests aren't working:

https://github.com/heroku/heroku-buildpack-nodejs/blob/master/bin/test#L202-L255

**Comment by hunterloftis** [comment]
Just made another test; `NPM_CONFIG_PRODUCTION` is behaving as expected. You might want to open a ticket (help.heroku.com) for the project you're talking about so we can debug that specifically.

```
hloftis():~/test/npmconfigprod (master)$ git push heroku master
Counting objects: 4, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (3/3), done.
Writing objects: 100% (4/4), 465 bytes | 0 bytes/s, done.
Total 4 (delta 0), reused 0 (delta 0)
remote: Compressing source files... done.
remote: Building source:
remote:
remote: -----> Node.js app detected
remote:
remote:        Node engine:         unspecified
remote:        Npm engine:          unspecified
remote:        Start mechanism:     none
remote:        node_modules source: package.json
remote:        node_modules cached: false
remote:
remote:        NPM_CONFIG_PRODUCTION=true
remote:        NODE_MODULES_CACHE=true
remote:
remote: -----> Installing binaries
remote:        Resolving node version (latest stable) via semver.io...
remote:        Downloading and installing node 0.10.35...
remote:
remote: -----> Building dependencies
remote:        No cache available
remote:        Installing node modules
remote:        npm WARN package.json npmconfigprod@1.0.0 No description
remote:        npm WARN package.json npmconfigprod@1.0.0 No repository field.
remote:        npm WARN package.json npmconfigprod@1.0.0 No README data
remote:        underscore@1.7.0 node_modules/underscore
remote:
remote: -----> Checking startup method
remote:
remote: -----> Finalizing build
remote:        Creating runtime environment
remote:        Exporting binary paths
remote:        Cleaning npm artifacts
remote:        Cleaning previous cache
remote:        Caching results for future builds
remote:
remote: -----> Build succeeded!
remote:
remote:        npmconfigprod@1.0.0 /tmp/build_3212747a2f5d737ce198b78c00cd7288
remote:        └── underscore@1.7.0
remote:
remote:        WARNING: Node version not specified in package.json
remote:        https://devcenter.heroku.com/articles/nodejs-support#specifying-a-node-js-version
remote:
remote:        WARNING: No Procfile, package.json start script, or server.js file found
remote:        https://devcenter.heroku.com/articles/nodejs-support#runtime-behavior
remote:
remote: -----> Discovering process types
remote:        Procfile declares types -> (none)
remote:
remote: -----> Compressing... done, 5.5MB
remote: -----> Launching... done, v3
remote:        https://tranquil-savannah-9081.herokuapp.com/ deployed to Heroku
remote:
remote: Verifying deploy... done.
To https://git.heroku.com/tranquil-savannah-9081.git
 * [new branch]      master -> master
hloftis():~/test/npmconfigprod (master)$ heroku config:set NPM_CONFIG_PRODUCTION=false
Setting config vars and restarting tranquil-savannah-9081... done, v4
NPM_CONFIG_PRODUCTION: false
hloftis():~/test/npmconfigprod (master)$ git commit --allow-empty -am 'empty' && git push heroku master
[master b919f1f] empty
Counting objects: 1, done.
Writing objects: 100% (1/1), 184 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 0 (delta 0)
remote: Compressing source files... done.
remote: Building source:
remote:
remote: -----> Node.js app detected
remote:
remote:        Node engine:         unspecified
remote:        Npm engine:          unspecified
remote:        Start mechanism:     none
remote:        node_modules source: package.json
remote:        node_modules cached: true
remote:
remote:        NPM_CONFIG_PRODUCTION=false
remote:        NODE_MODULES_CACHE=true
remote:
remote: -----> Installing binaries
remote:        Resolving node version (latest stable) via semver.io...
remote:        Downloading and installing node 0.10.35...
remote:
remote: -----> Building dependencies
remote:        Restoring node modules from cache
remote:        Pruning unused dependencies
remote:        npm WARN package.json npmconfigprod@1.0.0 No description
remote:        npm WARN package.json npmconfigprod@1.0.0 No repository field.
remote:        npm WARN package.json npmconfigprod@1.0.0 No README data
remote:        Installing any new modules
remote:        npm WARN package.json npmconfigprod@1.0.0 No description
remote:        npm WARN package.json npmconfigprod@1.0.0 No repository field.
remote:        npm WARN package.json npmconfigprod@1.0.0 No README data
remote:        lodash@2.4.1 node_modules/lodash
remote:
remote: -----> Checking startup method
remote:
remote: -----> Finalizing build
remote:        Creating runtime environment
remote:        Exporting binary paths
remote:        Cleaning npm artifacts
remote:        Cleaning previous cache
remote:        Caching results for future builds
remote:
remote: -----> Build succeeded!
remote:
remote:        npmconfigprod@1.0.0 /tmp/build_03f922040c935f74aba3bdd126549cdc
remote:        ├── lodash@2.4.1
remote:        └── underscore@1.7.0
remote:
remote:        WARNING: Node version not specified in package.json
remote:        https://devcenter.heroku.com/articles/nodejs-support#specifying-a-node-js-version
remote:
remote:        WARNING: No Procfile, package.json start script, or server.js file found
remote:        https://devcenter.heroku.com/articles/nodejs-support#runtime-behavior
remote:
remote: -----> Discovering process types
remote:        Procfile declares types -> (none)
remote:
remote: -----> Compressing... done, 5.7MB
remote: -----> Launching... done, v5
remote:        https://tranquil-savannah-9081.herokuapp.com/ deployed to Heroku
remote:
remote: Verifying deploy... done.
To https://git.heroku.com/tranquil-savannah-9081.git
   8a594a5..b919f1f  master -> master
hloftis():~/test/npmconfigprod (master)$ cat package.json
{
  "name": "npmconfigprod",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "underscore": "^1.7.0"
  },
  "devDependencies": {
    "lodash": "^2.4.1"
  }
}
```

**Comment by CharlieC3** [comment]
@hunterloftis 
Sure, thanks for the response. I'm trying to use this buildpack in conjunction with the deis/slugbuilder docker image. The github project for it is located here:
https://github.com/deis/deis/tree/master/builder/image/slugbuilder

This image allows the user to pass in environment variables when running a docker command. Mine looks something like this:
`docker run --interactive --attach stdin --attach stdout --attach stderr --env BUILDPACK_URL=https://github.com/heroku/heroku-buildpack-nodejs.git --env NODE_ENV=development --env NPM_CONFIG_PRODUCTION=false deis/slugbuilder`

So while the current version of this buildpack may support Heroku, I'm not sure if the current implementation caters to Deis or Flynn images as well.

Do you know if this should work? Passing in variables like I have above works for other buildpacks I've used in a similar manner.

**Comment by CharlieC3** [comment]
I tested my change locally and now the environment variables take the value I'm passing them from the terminal.
```
-----> Building dependencies
       No cache available
       Installing node modules
       npm WARN package.json redesign-static@0.0.29 No repository field.
       npm WARN cannot run in wd redesign-static@0.0.29 npm install -g grunt-cli (wd=/tmp/build)
       npm WARN deprecated grunt-lib-contrib@0.7.1: DEPRECATED. See readme: https://github.com/gruntjs/grunt-lib-contrib
       npm WARN cannot run in wd redesign-static@0.0.29 npm install && bundle install (wd=/tmp/build)
       grunt-shell@0.7.0 node_modules/grunt-shell
       └── chalk@0.4.0 (has-color@0.1.7, ansi-styles@1.0.0, strip-ansi@0.1.1)
```

Before the change, these devDependencies were being skipped despite the env variable being passed in.

**Comment by hunterloftis** [comment]
Ah I see - the buildpack API actually specifies that environment vars will come from the passed `ENV_DIR`:

- https://devcenter.heroku.com/articles/buildpack-api#bin-compile-summary

I'm not opposed to allowing env vars to come from *either* the environment *or* the `ENV_DIR`, but we need to make sure the official API (env dir) is always supported (and preferred in case of conflicting data with the env).

Something like what you had might work with the order changed:

```
export_env_dir $env_dir
export NPM_CONFIG_PRODUCTION=${NPM_CONFIG_PRODUCTION:-true}
export NODE_MODULES_CACHE=${NODE_MODULES_CACHE:-true}
```

We'll also need tests for the new behavior (accepting defaults from the environment as well as the env dir). Would you like to submit a PR for that? I'm happy to help.

**Pull request closed without merging**